### PR TITLE
WIP Jiterator reduction

### PR DESCRIPTION
**Pull request by ngimel**
This PR enables jit-compiled reductions and moves `prod` to be jit-compiled. 
Currently, only reductions that can use `func_wrapper` for automatic implementation of `reduce/project/translate_idx` opes are supported, there are a few TODOs for support of more complex reductions such as norms and max, that typically require full-fledged ReduceOps functor. Similarly, only reductions with a single input are supported.
Number of inputs is hardcoded to 1, which is true for our current reductions, but can be relaxed in the future.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
curious, will this line of work enable in the future jit-generation of economic, copy/upcast-avoiding variants to solve https://github.com/pytorch/pytorch/issues/55366 ?

**Comment by ngimel** [comment]
Yes, it will! We would still like to avoid jit-compiling the cross-product of dtype conversions, but we should be able to more easily accommodate some common cases like bool/uint8 -> int/int64

**Comment by vadimkantorov** [comment]
about cross-product: if it's jit-compiling, a given program rarely uses full cross-product, so it may be interesting to have a regime that indeed allows for jit-compiling everything (or more of things). even better if it can cache some compiled kernel artifacts in a given directory :)

**Comment by ngimel** [comment]
cc @kshitij12345, @zasdfgbnm

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -17,6 +18,9 @@
 #include <utility>
 #include <thrust/pair.h>
 
+#include <ATen/native/cuda/jit_utils.h>
+#include <iostream>
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/Reduce.cuh:22]: Classic

Comment (ngimel) [reply aten/src/ATen/native/cuda/Reduce.cuh:22]: It's a WIP :-)

**Comment by Lezcano** [comment]
Still wip? Otherwise I'll try to have a proper look at it on Monday

**Comment by ngimel** [comment]
No longer wip, tests are passing

**Review by mruberry — APPROVED** [review]
Awesome!

**Comment by seemethere** [comment]
@pytorchbot revert this, broke internal tests with:

```
aten/src\ATen/native/cuda/CUDAJitLoops.cuh(201): error C2398: Element '1': conversion from 'int' to 'unsigned int' requires a narrowing conversion
```

[Internal link](https://www.internalfb.com/diff/D35337550?selected_signal=bGFuZF9ibG9ja2luZ19sYW5kX2lzc3VlOmMyRnVaR05oYzNSc1pWOTNiM0pyWm14dmQxOXlkVzQ2TnpJd05UYzFPVFF3TkRNd01UTTJOVEk9OjpTTFBDSURpZmZUaW1lVmFsaWRhdGlvbjo6MzY4MjMzMzU1MTk4ODc3&selected_signal_verification_phase=2&dst_version_fbid=368233355198877)

**Pull request closed without merging**